src/rgw/rgw_rest_sts.cc

```diff
@@ -339,13 +338,262 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
+{
+  struct hostent* host = gethostbyname(hostname.c_str());
+  if (!host) {
+    ldpp_dout(dpp, 0) << "gethostbyname failed for host: " << hostname << dendl;
+    return -1;
+  }
+
+  struct sockaddr_in addr;
+  memset(&addr, 0, sizeof(addr));
+  addr.sin_family = AF_INET;
+  addr.sin_port = htons(port);
+  memcpy(&addr.sin_addr, host->h_addr, host->h_length);
+
+  int sock = socket(AF_INET, SOCK_STREAM, 0);
+  if (sock < 0) {
+    ldpp_dout(dpp, 0) << "creation of socket failed: " << sock << dendl;
+    return -1;
+  }
+
+  int ret = connect(sock, (struct sockaddr*)&addr, sizeof(addr));
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "connection to socket failed: " << ret << dendl;
+    close(sock);
+    return -1;
+  }
+
+  return sock;
+}
+
+std::string
+WebTokenEngine::extract_last_certificate(const DoutPrefixProvider* dpp, const std::string& pem_chain) const
+{
+  const std::string BEGIN_MARKER = "-----BEGIN CERTIFICATE-----";
+  const std::string END_MARKER = "-----END CERTIFICATE-----";
+
+  // Find the last occurrence of BEGIN marker
+  size_t begin_pos = pem_chain.rfind(BEGIN_MARKER);
+  if (begin_pos == std::string::npos) {
+    ldpp_dout(dpp, 0) << "No BEGIN marker found in certificate chain" << dendl;
+    throw std::runtime_error("No BEGIN marker found in certificate chain");
+  }
+
+  // Find the END marker that comes after the last BEGIN marker
+  size_t end_pos = pem_chain.find(END_MARKER, begin_pos);
+  if (end_pos == std::string::npos) {
+    ldpp_dout(dpp, 0) << "No matching END marker found after last BEGIN marker" << dendl;
+    throw std::runtime_error("No matching END marker found after last BEGIN marker");
+  }
+
+  // Calculate the start and length of the complete certificate (including markers)
+  size_t cert_length = (end_pos + END_MARKER.length()) - begin_pos;
+
+  // Extract the complete certificate
+  std::string last_cert = pem_chain.substr(begin_pos, cert_length);
+
+  return last_cert;
+}
+
 void
-WebTokenEngine::validate_signature(const DoutPrefixProvider* dpp, const jwt::decoded_jwt& decoded, const string& algorithm, const string& iss, const vector<string>& thumbprints, optional_yield y) const
+WebTokenEngine::shutdown_ssl(const DoutPrefixProvider* dpp, SSL* ssl, SSL_CTX* ctx) const
+{
+  int status = SSL_shutdown(ssl);
+  //status = 0, we have issued shutdown but not acknowledged by remote connection
+  //status = 1, remote connection has shutdown
+  //status !=1 && != 0, error
+  if (status == 0) {
+    status = SSL_shutdown(ssl);
+    if (status != 1) {
+      auto error = SSL_get_error(ssl, status);
+      //do we need to notify end user of ssl shutdown failure?
+      ldpp_dout(dpp, 10) << "SSL shutdown failed with error: "<< error << dendl;
+    }
+  } else if (status != 1) {
+    auto error = SSL_get_error(ssl, status);
+    ldpp_dout(dpp, 10) << "SSL shutdown failed with error: "<< error << dendl;
+  }
```

Comment:
the `(status != 1)` blocks are duplicated twice. can we simplify be removing the `else`?
```c++
if (status == 0) {
  status = SSL_shutdown(ssl);
}
if (status != 1) {
  ...
```
otherwise the `shutdown_ssl()` stuff looks good 👍